core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -2519,7 +2520,13 @@ private void loadMetadataAndJournal(RpcContext rpcContext, LockedInodePath inode
         InodeDirectory inode = (InodeDirectory) inodePath.getInode();
 
         if (options.getLoadDescendantType() != DescendantType.NONE) {
-          UfsStatus[] children = ufs.listStatus(ufsUri.toString());
+          ListOptions listOptions = ListOptions.defaults();
+          if (options.getLoadDescendantType() == DescendantType.ALL) {
+            listOptions.setRecursive(true);
+          } else {
+            listOptions.setRecursive(false);
+          }
+          UfsStatus[] children = ufs.listStatus(ufsUri.toString(), listOptions);
```

Comment:
I don't understand how this works. This will recursively list from UFS, but later, there will be a recursive call again to `loadMetadataAndJournal` (existing behavior). Shouldn't the listing returned by the recursive UFS listing be processed fully, and no more recursive calls would be necessary to `loadMetadataAndJournal`?

It seems like the existing code will call the UFS recursive list for every descendant directory under the input path?

Reply:
The recursive call to `loadMetadataAndJournal` is always going to have `DescendantType.NONE`. The logic behind this code is, we grab a list of `ufsStatus` from the ufs, and we go through each one of them and call either `loadFileMetadataAndJournal` or `loadDirectoryMetadataAndJournal` depending on whether that ufsStatus is a file or directory. Since we always pass in a `ufsStatus`, these calls would not be calling the ufs for more information.